File: simulators/typescript/src/scenarios/connectionStorms.ts
import { Client } from "pg";
import { Config, RawResult } from "../types";

export async function connectionStorms(config: Config): Promise<RawResult> {
  let ops = 0, errors = 0;
  const latencies: number[] = [];

  const start = Date.now();
  const storms = Array.from({ length: config.concurrency * 2 }, () => (async () => {
    const t = performance.now();
    const client = new Client({ host: config.host, port: config.port, user: config.user, password: config.password, database: config.database });
    client.on("error", () => {}); // suppress unhandled error events
    try {
      await client.connect();
      await client.query("SELECT 1");
      ops++;
    } catch { errors++; ops++; }
    finally { try { await client.end(); } catch {} }
    latencies.push(performance.now() - t);
  })());

  await Promise.all(storms);
  return { totalOps: ops, errors, durationMs: Date.now() - start, latencies };
}

File: simulators/go/scenarios/connection_storms.go
package scenarios

import (
	"context"
	"sync"
	"sync/atomic"
	"time"
)

// ConnectionStorms simulates rapid connect/disconnect cycles — the worst case for a connection pooler.
// Each goroutine opens a fresh connection, runs one query, and closes immediately.
// The storm intentionally spawns 2× the configured concurrency to exceed the backend pool size,
// so capacity rejections (SQLSTATE 53300) are expected and filtered out.
//
// In duration mode each goroutine keeps storming (connect/query/close) until the deadline expires.
// In iteration mode each goroutine performs a single storm cycle (original behaviour).
func ConnectionStorms(ctx context.Context, cfg *Config) (*Result, error) {
	var (
		ops       atomic.Int64
		errCount  atomic.Int64
		mu        sync.Mutex
		latencies []time.Duration
	)
	errs := NewErrorSampler(10)

	start := time.Now()
	var wg sync.WaitGroup

	storms := cfg.Concurrency * 2
	for i := 0; i < storms; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			for iter := 0; cfg.Continue(ctx, iter, 1); iter++ {
				t := time.Now()
				conn, err := ConnectClient(ctx, cfg)
				if err != nil {
					// Capacity rejections are expected when exceeding the backend pool size.
					if !IsCapacityError(err) {
						if ctx.Err() != nil {
							return
						}
						errCount.Add(1)
						errs.Record(err)
					}
					ops.Add(1)
					continue
				}

				// conn.Exec with no parameters uses the simple protocol (single Query message),
				// bypassing the extended query protocol entirely — correct in both pool modes.
				_, err = conn.Exec(ctx, "SELECT 1")
				ops.Add(1)
				if err != nil {
					if !IsCapacityError(err) {
						if ctx.Err() == nil && !conn.IsClosed() {
							errCount.Add(1)
							errs.Record(err)
						}
					}
				}

				_ = conn.Close(context.Background())
				d := time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
			}
		}()
	}

	wg.Wait()
	duration := time.Since(start)

	return &Result{
		TotalOps:  int(ops.Load()),
		Errors:    int(errCount.Load()),
		Duration:  duration,
		Latencies: latencies,
		Notes:     errs.Notes(),
	}, nil
}
